Chat-action-bar: Use parsed markdown to compose a message

just-carlod opened the pull request:
Added markdown parsing functionality 

https://user-images.githubusercontent.com/58047802/173435858-e9372a66-06a2-4a1f-b866-e7add9b336c5.mp4

marhkb commented on src/session/content/chat_action_bar.rs:
```diff
@@ -560,3 +566,15 @@ fn message_bar_visibility_in_groups(
+        let FormattedText::FormattedText(text) = text;
+        return Some(text);
+    }
+    None
```
Nitpick: Try to use combinators. These are much more elegant than this 80s if-else-return C-style.

```suggestion
    functions::parse_markdown(formatted_text, client_id)
        .await
        .map(|text| {
            let FormattedText::FormattedText(text) = text;
            text
        })
        .ok()
```

just-carlod replied:
> Nitpick: Try to use combinators. These are much more elegant than this 80s if-else-return C-style.

Thanks for your advice

marhkb commented on src/session/content/chat_action_bar.rs:
```diff
@@ -560,3 +566,16 @@ fn message_bar_visibility_in_groups(
     .upcast()
 }
+
+async fn parse_markdown(
```
I think @melix99 has already mentioned it. The markdown parsing now only works in the `ChatActionBar` but not in the `SendPhotoDialog`. So the parsing should happen directly in the `MessageEntry` to make it reusable.

just-carlod replied:
Ok, I'll do it tomorow, For this message entry will need `WeakRef<Session>` to get `client_id()`

just-carlod replied:
Well, I've been thinking about this problem and I think passing `Session` or `client_id` to the MessageEntry contructor would conflict with `Default` implementation
```rust 

impl Default for MessageEntry {
    fn default() -> Self {
        Self::new()
    }
}
```
I'm not sure if it's worth it to modify the constructor if there is a simpler workaround and the text is being parsed only in two places (ChatActionBar, SendPhotoDialog) which have already access to `client_id()`.

marhkb replied:
`ChatActionBar`, which contains the `MessageEntry`, also contains the `Chat`. `Chat` contains the `Session`. `Session` contains the `client_id`.

`SendPhoto` also contains both the `Chat` and a `MessageEntry`

So you also need to add a glib GObject property "chat" to `MessageEntry` and bind that to the `chat` property of the parent widget (leave the Default impl in place).

https://github.com/melix99/telegrand/blob/f9a5bd88162211846d392f8db72a3bd7aff90220/data/resources/ui/content-chat-action-bar.ui#L21

https://github.com/melix99/telegrand/blob/f9a5bd88162211846d392f8db72a3bd7aff90220/data/resources/ui/content-send-photo-dialog.ui#L33

If you have implemented the property, you can just bind it.
```xml
                <child>
                  <object class="MessageEntry" id="caption_entry">
                    <property name="hexpand">True</property>
                    <property name="placeholder-text" translatable="yes">Caption</property>
                    <binding name="chat">
                      <lookup name="chat">ContentSendPhotoDialog</lookup>
                    </binding>
                  </object>
                </child>
```

just-carlod commented on src/session/content/send_photo_dialog.rs:
```diff
@@ -134,6 +134,8 @@ impl SendPhotoDialog {
         let paintable = imp.picture.paintable().unwrap();
         let width = paintable.intrinsic_width();
         let height = paintable.intrinsic_height();
+        imp.caption_entry.set_chat(Some(chat.to_owned()));
```
I know it's a hacky way, but it does work unlike binding chat in a template file :/
![screenshot](https://user-images.githubusercontent.com/58047802/187710523-bb54d8d6-48ea-45d3-8e44-ae686f57adbe.png)

marhkb commented on src/session/content/send_photo_dialog.rs:
```diff
@@ -100,12 +134,15 @@ impl SendPhotoDialog {
 
+    pub(crate) fn chat(&self) -> Option<&Chat> {
+        self.imp().chat.get()
+    }
```
Either remove this method or reuse it in those places where currently `self.imp().chat.get()` is used.

marhkb commented on src/session/components/message_entry.rs:
```diff
@@ -216,6 +228,40 @@ impl MessageEntry {
 
+    /* formatted_text is being replaced every time text buffer has changed,
+     * so it's very unefficient to parse markdown text when it's not nesessary,
+     * instead we should use a separate method, that will be called only when composing a message */
```
What does this mean? Isn't this parsing function just called when we send a message or save a draft message? I think this comment can be removed.

just-carlod replied:
That was an argumentation why to create `MessageEntry::parsed_formatted_markdown()`, but you're right it was useless

marhkb commented on src/session/components/message_entry.rs:
```diff
@@ -216,6 +228,40 @@ impl MessageEntry {
+                let EnumFormattedText::FormattedText(text) = text;
+                text
+            })
+            .ok()
```
I would change the body a bit
```suggestion
        let text = self.imp().formatted_text.borrow().clone().map(|f| f.0)?;
        let client_id = self.chat().unwrap().session().client_id();

        functions::parse_markdown(text.clone(), client_id)
            .await
            .map(|text| {
                let EnumFormattedText::FormattedText(text) = text;
                text
            })
            .ok()
            .or(Some(text))
```
The points the changes addresses:
1. An `?` is used to return early
2. We can use `Self::chat()` method, which you introduced
3. We don't need to create a let binding for chat and clone it
4. Not sure here, but if parsing has failed, we could at least return the original text.

just-carlod replied:
Thanks for your help, didn't know`or` function existed

marhkb commented on src/session/components/message_entry.rs:
```diff
@@ -16,6 +18,7 @@ mod imp {
     #[derive(Debug, Default, CompositeTemplate)]
     #[template(resource = "/com/github/melix99/telegrand/ui/components-message-entry.ui")]
     pub(crate) struct MessageEntry {
+        pub(crate) chat: RefCell<Option<Chat>>,
```
I recommend using a `WeakRef` here and make it `pub(super)`
```suggestion
        pub(super) chat: WeakRef<Chat>,
```

marhkb commented on src/session/components/message_entry.rs:
```diff
@@ -216,6 +228,40 @@ impl MessageEntry {
+
+    pub(crate) fn chat(&self) -> Option<Chat> {
+        self.imp().chat.borrow().clone()
+    }
```
If you use a `WeakRef`.

```suggestion
    pub(crate) fn chat(&self) -> Option<Chat> {
        self.imp().chat.upgrade()
    }
```

marhkb commented on src/session/components/message_entry.rs:
```diff
@@ -216,6 +228,40 @@ impl MessageEntry {
+    pub(crate) fn set_chat(&self, chat: Option<Chat>) {
+        if self.chat() != chat && chat.is_some() {
+            self.imp().chat.replace(chat);
+        }
```
1. I would keep the scheme/style we use to set properties
2. Testing for `Some` is pointless
3. You forgot to notify the property

```suggestion
        if self.chat().as_ref() == chat {
            return;
        }
        self.imp().chat.set(chat);
        self.notify("chat");
```

marhkb commented on src/session/components/message_entry.rs:
```diff
@@ -216,6 +229,33 @@ impl MessageEntry {
         self.imp().text_view.buffer().set_text(&text);
     }
 
+    pub(crate) async fn parsed_formatted_text(&self) -> Option<FormattedText> {
```
Maybe we should call this function just `as_markdown` to give it a more precise name.

marhkb commented on src/session/content/send_photo_dialog.rs:
```diff
@@ -56,6 +57,39 @@ mod imp {
+
+        fn property(&self, obj: &Self::Type, _id: usize, pspec: &glib::ParamSpec) -> glib::Value {
+            match pspec.name() {
+                "chat" => obj.imp().chat.get().to_value(),
```
```suggestion
                "chat" => self.chat.get().to_value(),
```

marhkb commented on src/session/components/message_entry.rs:
```diff
@@ -76,6 +80,13 @@ mod imp {
+                    glib::ParamSpecObject::new(
+                        "chat",
+                        "Chat",
+                        "The chat on which text is typed",
```
```suggestion
                        "The chat in which the messages will be sent",
```

just-carlod committed:
chat-action-bar: Use parsed markdown to compose a message

Added function to parse markdown text which can be used then to compose a message

Co-authored-by: Marcus Behrendt <marcus.behrendt.86@gmail.com>
